### PR TITLE
group email fix and some code cleanup

### PR DESCRIPTION
**Pull request by gymad**
## Description

- Clean up the code of class ListViewDataEmails, split the critical getListViewData() function to separated smaller units.
- Clean up the php error log and fix these and/or shows the possible errors in crm log (ref for e.g: #3483)
- Email Settings doesn't shows the folders in deleted Inbound Email Accounts
- Email Settings shows group account too
- PR also contains a fix: when hard delete a folder it's delete the sub-folders also properly
- PR added some warning message into crm log when unexpected error happens without handling
- bugfix: if settings popup did not used the clear user email preferences
- bugfix: do not shows create link if no module for listview
- bugfix: Email list-view shows group IE as default

## Motivation and Context

ListViewDataEmails::getListViewData() method too long, impossible to debugging in a more than 1000 line complex function, adding functional tests close to impossible.

## How To Test This

(rebuild group js first)
- PHP error.log should be clean

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 1]
```diff
@@ -44,15 +44,31 @@
...
 class ListViewDataEmails extends ListViewData
 {
+
+
+    /**
+     * enum('crm', 'imap')
+     *
+     * @var string
+     */
+    protected $searchType;
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: We should implement an abstract class with constants. This simplifies the type in the php doc.

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 2]
```diff
@@ -44,15 +44,31 @@
...
+
     /**
      * @var array
-     * when searching the imap filter map the crm fields
-     * to the imap fields.
+     * when searching the IMap filter map the crm fields
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: IMap? should be IMAP?

IMAP === (Internet Message Access Protocol)

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 3]
```diff
@@ -116,849 +132,590 @@ public function __construct()
...
-        $params = array(),
-        $id_field = 'id',
-        $singleSelect = true,
-        $id = null
-    ) {
-        global $current_user;
-        global $sugar_config;
-        global $db;
-        global $mod_strings;
+    protected function getInboundEmail(User $currentUser, Folder $folder) {
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: This is just a note for the pull request that type hinting with a callable is supported in PHP 5.4 and above. No action is required here.

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 4]
```diff
@@ -116,849 +132,590 @@ public function __construct()
...
-                    if(empty($_REQUEST['action']) || $_REQUEST['action'] != 'Popup'){
-                        $_SESSION['MAILMERGE_MODULE'] = $seed->module_dir;
-                    }
+    /**
+     * get email header status
+     *
+     * possible return variables:
+     * enum('replied', 'draft', 'read', 'unread')
+     *
+     * @param array $emailHeader
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: We should use abstract class with const properties instead

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 5]
```diff
@@ -116,849 +132,590 @@ public function __construct()
...
-                    $prevOffset = -1;
-                    $endOffset = 0;
 
+                    $limit = $sugar_config['list_max_entries_per_page'];
 
-                    if ($total > $limitPerPage) {
-                        $nextOffset = $offset + $limitPerPage;
-                    }
+                    $search = new ListViewDataEmailsSearchOnCrm($this);
+                    $ret = $search->search($filter_fields, $request, $where, $inboundEmail, $params, $seed, $singleSelect, $id, $limit, $current_user, $id_field, $offset);
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: Line goes over the 120 character limit. Consider splitting the into on line per argument.

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 6]
```diff
@@ -116,849 +132,590 @@ public function __construct()
...
-                    if ($total > 0) {
-                        $endOffset = $total / $limitPerPage;
-                    }
 
-                    $pageData['offsets']['current'] = $offset;
-                    $pageData['offsets']['total'] = $total;
-                    $pageData['offsets']['next'] = $nextOffset;
-                    $pageData['offsets']['prev'] = $prevOffset;
-                    $pageData['offsets']['end'] = ceil($endOffset);
+                    $limitPerPage = isset($sugar_config['list_max_entries_per_page']) && (int)$sugar_config['list_max_entries_per_page'] ? $sugar_config['list_max_entries_per_page'] : 10;
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: Lots of white space. Perhaps it would be easier to read with out the extra double lines.

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+    /**
+     * @param array $filterFields
+     * @param array $request $_REQUEST
+     * @param string $where
+     * @param InboundEmail $inboundEmail
+     * @param array $params
+     * @param Email $seed
+     * @param bool $singleSelect
+     * @param ??? $id
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Should type be string?

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+     * @param int $limit
+     * @param User $currentUser
+     * @param string $idField
+     * @param int $offset
+     * @return array
+     */
+    public function search($filterFields, $request, $where, InboundEmail $inboundEmail, $params, Email $seed, $singleSelect, $id, $limit, User $currentUser, $idField, $offset) {
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Too long

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+
+
+        $this->lvde->setVariableName($seed->object_name, $crmWhere, $this->lvde->listviewName, $id);
+
+        SugarVCR::erase($seed->module_dir);
+        $this->lvde->seed =& $seed;
+        $totalCounted = empty($GLOBALS['sugar_config']['disable_count_query']);
+        $_SESSION['MAILMERGE_MODULE_FROM_LISTVIEW'] = $seed->module_dir;
+        if(empty($request['action']) || $request['action'] != 'Popup'){
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Can safely use !==

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        $data = array();
+
+        $rows = array();
+        $count = 0;
+        $idIndex = array();
+        $id_list = '';
+
+        $crmEmails = $this->lvde->db->query($crmEmailsQuery);
+
+        while(($row = $this->lvde->db->fetchByAssoc($crmEmails)) != null)
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Safely use !==

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+            $count++;
+        }
+
+        if (!empty($id_list))
+        {
+            $id_list = '('.substr($id_list, 1).')';
+        }
+
+        SugarVCR::store($this->lvde->seed->module_dir,  $crmEmailsQuery);
+        if($count != 0) {
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Safely use !==

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+                $temp = clone $seed;
+                $dataIndex = count($data);
+
+                $temp->setupCustomFields($temp->module_dir);
+                $temp->loadFromRow($row);
+                if (empty($this->lvde->seed->assigned_user_id) && !empty($temp->assigned_user_id)) {
+                    $this->lvde->seed->assigned_user_id = $temp->assigned_user_id;
+                }
+                if($idIndex[$row[$idField]][0] == $dataIndex){
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Safely use ===

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+        }
+        else if (isset($request["searchFormTab"]) && $request["searchFormTab"] == "basic_search")
+        {
+            // TODO: figure out what was the SearchFormReports???
+            if($seed->module_dir == "Reports") $searchMetaData = SearchFormReports::retrieveReportsSearchDefs();
+            else $searchMetaData = SearchForm::retrieveSearchDefs($seed->module_dir); // TODO: figure out which SearchForm is it?
+
+            $basicSearchFields = array();
+
+            if( isset($searchMetaData['searchdefs']) && isset($searchMetaData['searchdefs'][$seed->module_dir]['layout']['basic_search']) )
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: PSR-2: please use {} for if statements

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php [hunk 14]
```diff
@@ -0,0 +1,316 @@
+<?php
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php]: The name ListViewDataEmailsSearchOnIMap seems strange because IMAP is an acronym.

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+     * @param Email $seed
+     * @param string $where
+     * @param string $id
+     * @param InboundEmail $inboundEmail
+     * @param array $filter
+     * @param Folder $folderObj
+     * @param User $currentUser
+     * @param string $folder
+     * @param int $limit
+     * @param ??? $limitPerPage
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php]: type === string?

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+     * @return array
+     */
+    public function search(Email $seed, $where, $id, InboundEmail $inboundEmail, $filter, Folder $folderObj, User $currentUser, $folder, $limit, $limitPerPage) {
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php]: Too long

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 17]
```diff
@@ -44,15 +44,31 @@
...
+include_once 'modules/Emails/Folder.php';
+include_once 'modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php';
+include_once 'modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php';
+
+
+
 class ListViewDataEmails extends ListViewData
 {
+
+
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: Missing $seed property and documentation see modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php: line 67

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+        $pageData['stamp'] = $this->lvde->stamp;
+        $pageData['access'] = array('view' => $this->lvde->seed->ACLAccess('DetailView'), 'edit' => $this->lvde->seed->ACLAccess('EditView'));
+        $pageData['idIndex'] = $idIndex;
+        if(!$this->lvde->seed->ACLAccess('ListView')) {
+            $pageData['error'] = 'ACL restricted access';
+        }
+
+        $queryString = '';
+
+        if( isset($request["searchFormTab"]) && $request["searchFormTab"] == "advanced_search" ||
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: Code inspection says that the conditions are confusing. You need to add ( condition ) instead.

File: include/SugarFolders/SugarFolders.php [hunk 19]
```diff
@@ -682,6 +687,34 @@ public function getFoldersForSettings($focusUser = null)
...
+     * @param array $folders - array of folders table rows
+     * @return array
+     */
+    private function removeDeletedFolders($folders) {
+
+        $ret = array();
+
+        foreach($folders as $folder) {
+            $correct = false;
+            if(!$folder['id']) {
```
Comment (gymad) [inline include/SugarFolders/SugarFolders.php]: > Why are we using $correct twice? what is correct? The id?

No, correct each one which are not deleted. See the original function. Only different is: I need the non-deleted folders, so each one is correct if no folder[id] and the IE account exists.

File: include/TemplateHandler/TemplateHandler.php [hunk 20]
```diff
@@ -434,8 +434,8 @@ public function createQuickSearchCode($defs, $defs2, $view = '', $module = '')
 
                 if ($field['type'] === 'relate' &&
                     isset($field['module']) &&
-                    !empty($field['quicksearch']) ||
-                    preg_match('/_name$|_c$/si', $name)
+                    (!empty($field['quicksearch']) ||
```
Comment (gymad) [inline include/TemplateHandler/TemplateHandler.php]: extra line removed.

File: modules/Emails/EmailUIAjax.php [hunk 21]
```diff
@@ -42,6 +42,45 @@
     die('Not A Valid Entry Point');
 }
 
+
+/**
+ * handle requested subscriptions
+ *
+ * @param array $subs
+ * @param $email
```
Comment (gymad) [inline modules/Emails/EmailUIAjax.php]: added

File: modules/Emails/include/ListView/ListViewDataEmails.php (hunk 17 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: php doc added
(class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmails.php (hunk 1 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: abstraction are there: ListViewDataEmailsSearchAbstract
(class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmails.php (hunk 2 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: PhpStorm shows misspelling for "imap". "IMap" acceptable.

File: modules/Emails/include/ListView/ListViewDataEmails.php (hunk 3 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: from version 7.8.x the minimum supported PHP version is 5.5+

File: modules/Emails/include/ListView/ListViewDataEmails.php (hunk 4 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: I aggree

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php (hunk 8 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: yes, I know :D

Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php (hunk 9 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php (hunk 10 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php (hunk 11 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php (hunk 12 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php (hunk 18 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnCrm.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php (hunk 14 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php (hunk 15 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)

File: modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php (hunk 16 above)
Comment (gymad) [inline modules/Emails/include/ListView/ListViewDataEmailsSearchOnIMap.php]: (class ListViewDataEmailsSearchOnIMap and class ListViewDataEmailsSearchOnCrm not completely refactored yet)
